### ISSUE TITLE
Not converting str to Literal despite proper type checking

### ISSUE THREAD

**Issue by bfelbo** [microsoft/pyright#615]
**Describe the bug**
See the code below. As the assert ensures the str is complying with the type, pyright should not report any issues.

**To Reproduce**
```
from typing_extensions import Literal

allowed_types = set(["Male", "Female"])
Gender = Literal["Male", "Female"]

def print_gender(gender: Gender):
    print(gender)

# Unknown string as it's been retrieved from elsewhere
strRetrievedFromDB = "x"  # type: str

assert strRetrievedFromDB in allowed_types
print_gender(strRetrievedFromDB)
```

**Expected behavior**
No pyright issues.

**VS Code extension or command-line**
VS code extension 1.1.31

**Comment by erictraut** [comment]
A type checker can't assume that `assert strRetrievedFromDB in allowed_types` means that `strRetrievedFromDB` is of type `Literal["Male", "Female"]`. You happen to know that's how set iteration and str comparison works, but a type checker can't make those assumptions. It needs to work from the type information provided in builtins.pyi, etc.

**Comment by bfelbo** [comment]
I see. Thanks.

**Comment by LouisStAmour** [comment]
I can’t say whether the problem lies within Python or the type checkers, but it strikes me that this kind of assertion is exactly the sort of thing TypeScript is slowly working itself towards, if it doesn’t already support it. TypeScript’s support is still under active development, for example https://github.com/microsoft/TypeScript/issues/37241 but my point is that eventually, type-checking should be able to infer narrowing of types in even more complex examples involving literals or finals. As newer ways to express constraints in the type system or assert them in Python are added or noticed, obviously this will be a bit of a never-ending task to reduce the number of explicit annotations of types required, while correctly inferring types at any point in a program...

**Comment by LouisStAmour** [comment]
For example, if the set were final, and thus a set containing a literal union of types “Male”, “Female”, “Other” then I would assume the original code would indeed have no errors found.
